### PR TITLE
Extract and simplify level manager into LSM manager

### PR DESCRIPTION
**Pull request by purplefox**
This PR extracts and simplifies the level manager into a new lsm.Manager struct.

It includes a more efficient way of representing table entries in memory, whereby the table entries aren't serialized/deserialized when the master record is, but kept in their []byte form. This allows quicker serialization/deserialization and a lower memory footprint. Both of these are important if we're going to be store metadata in S3 on each update.

### CODE REVIEW

File: lsm/compaction_worker.go [hunk 1]
```diff
@@ -0,0 +1,530 @@
...
+	}
+	mergeDur := time.Now().Sub(mergeStart)
+	log.Debugf("merge for job %s took %d ms", job.id, mergeDur.Milliseconds())
+	// Now push the tables to the cloud store
+	var ids []sst.SSTableID
+	for _, info := range infos {
+		id := []byte(fmt.Sprintf("sst-%s", uuid.New().String()))
+		log.Debugf("compaction job %s created sstable %v", job.id, id)
+		ids = append(ids, id)
+		for {
```
Comment (stanislavkozlovski) [inline lsm/compaction_worker.go:230]: would it make sense to have a timeout here?

File: lsm/structs.go [hunk 2]
```diff
@@ -0,0 +1,639 @@
...
+	u, offset = encoding.ReadUint64FromBufferLE(buff, offset)
+	l.Bytes = int(u)
+	u, offset = encoding.ReadUint64FromBufferLE(buff, offset)
+	l.Entries = int(u)
+	u, offset = encoding.ReadUint64FromBufferLE(buff, offset)
+	l.Tables = int(u)
+	return offset
+}
+
+type MasterRecord struct {
```
Comment (stanislavkozlovski) [inline lsm/structs.go:344]: to confirm my base level architecture understanding - this is the same for every agent in the distributed system, right? Its source of truth is S3. (this is how I see the old `levels/structs.go` master record get created

File: lsm/compaction.go [hunk 3]
```diff
@@ -0,0 +1,852 @@
...
+func (lr *lockedRange) overlaps(rng *lockedRange) bool {
+	dontOverlapRight := bytes.Compare(rng.start, lr.end) > 0
+	dontOverlapLeft := bytes.Compare(rng.end, lr.start) < 0
+	dontOverlap := dontOverlapLeft || dontOverlapRight
+	return !dontOverlap
+}
+
+func (lm *Manager) maybeScheduleCompaction() error {
+	// Get a level to compact (if any)
+	level, numTables := lm.chooseLevelToCompact()
```
Comment (stanislavkozlovski) [inline lsm/compaction.go:59]: high level architecture q - Could two agents independently be compacting the same level at once? Would that waste cycles?

afaict yes. would this be avoidable with an s3 conditional write?

File: lsm/validate.go [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+
+import (
+	"bytes"
+	"github.com/spirit-labs/tektite/asl/errwrap"
+	log "github.com/spirit-labs/tektite/logger"
+	"github.com/spirit-labs/tektite/objstore"
+	"github.com/spirit-labs/tektite/sst"
+)
+
+// Validate checks the LSM is sound - no overlapping keys in L > 0 etc
```
Comment (stanislavkozlovski) [inline lsm/validate.go:11]:
```suggestion
// Validate checks the LSM is sound - no overlapping keys in L > 0, keys in each table are sorted, etc.
```

File: lsm/compaction_worker.go (hunk 1 above)
Comment (purplefox) [reply lsm/compaction_worker.go:230]: The loop here only retries if there is some kind of unavailability of the object store - e.g. temporary network problem or S3 is down. Usually for these kinds of things we just want to keep retrying until it comes back up again. Consider what else we could do? If we timed out, after say, 10 minutes or a day, what then? The agent is useless if it cannot contact the object store, so  I guess we could terminate the process. But what if the network issue is resolved? Now we have restart the process. I think the best thing we can do in these kinds of situations is effectively have an infinite timeout.

File: lsm/structs.go (hunk 2 above)
Comment (purplefox) [reply lsm/structs.go:344]: There would be one master record for each instance of the LSM, and then we shard the overall database by having multiple LSM instances, each covering different key ranges. These instances live on different agents in the cluster to distribute the load more evenly. This is the source of truth for the LSM instance and records which SSTables are in which levels of the LSM (amongst other things), so it holds the structure of the LSM.

File: lsm/compaction.go (hunk 3 above)
Comment (purplefox) [reply lsm/compaction.go:59]: Compaction only occurs on the agent that has the LSM controller for the shard.

**Pull request merged**